typing/rec_check.ml

```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
         Dynamic
   in classify_expression Ident.empty
 
-let rec expression : Env.env -> Typedtree.expression -> Use.t =
-  fun env exp -> match exp.exp_desc with
+let remove_pat pat env =
+  Env.remove_list (pat_bound_idents pat) env
+
+let remove_patlist pats env =
+  List.fold_right remove_pat pats env
+
+(* Usage mode judgments.
+
+   There are two main groups of judgment functions:
+
+   - Judgments of the form "Gamma |- ... : m"
+     compute the environment of a subterm from its mode,
+     so the corresponding function has type [... -> mode -> Env.t].
+
+     We write ['a -> term_judg] in this case.
+
+   - Judgments of the form "Gamma |- ... : m -| Gamma'"
+     correspond to binding constructs that have both an
+     exterior environment (the environment of the whole term)
+     and an interior environment (the environment after the binding
+     construct has introduced new names in scope).
+
+     For example, a toplevel let-binding could be given
+     the following rule:
+
+       G |- e : m + m'
+       -------------------------------
+       G |- (let x = e : m) -| x:m', G
+
+     In this case, the corresponding function has type
+     [... -> mode -> Env.t -> Env.t]: it takes the mode
+     and interior environment as inputs, and returns
+     the exterior environment.
+
+     We write ['a -> bind_judg] in this case.
+*)
+type term_judg = mode -> Env.t
+type bind_judg = mode -> Env.t -> Env.t
+
+let option : 'a. ('a -> term_judg) -> 'a option -> term_judg =
+  fun f o m -> match o with
+    | None -> Env.empty
+    | Some v -> f v m
+let list : 'a. ('a -> term_judg) -> 'a list -> term_judg =
+  fun f li m ->
+    List.fold_left (fun env item -> Env.join env (f item m)) Env.empty li
+let array : 'a. ('a -> term_judg) -> 'a array -> term_judg =
+  fun f ar m ->
+    Array.fold_left (fun env item -> Env.join env (f item m)) Env.empty ar
+
+let single : Ident.t -> term_judg = Env.single
+let remove_id : Ident.t -> term_judg -> term_judg =
+  fun id f m -> Env.remove id (f m)
+let remove_ids : Ident.t list -> term_judg -> term_judg =
+  fun ids f m -> Env.remove_list ids (f m)
+
+let join : term_judg list -> term_judg =
+  fun li m -> Env.join_list (List.map (fun f -> f m) li)
+
+let empty = fun _ -> Env.empty
+
+(* A judgment [judg] takes a mode from the context as input, and
+   returns an environment. The judgment [judg << m], given a mode [m']
+   from the context, evaluates [judg] in the composed mode [m'[m]]. *)
+let (<<) : term_judg -> mode -> term_judg =
+  fun f inner_mode -> fun outer_mode -> f (compos outer_mode inner_mode)
+
+(* A binding judgment [binder] expects a mode and an outer environment,
+   and returns an inner environment. [binder >> judg] computes
```

Comment:
This is in contradiction with the comment above? (which says that `bind_judg` takes an interior environment and returns an exterior environment)